app/src/main/java/com/nervousfish/nervousfish/data_objects/Contact.java

```diff
@@ -38,29 +40,56 @@ public Contact(final String name, final IKey key) {
      */
     public Contact(final String name, final Collection<IKey> keys) {
         this.name = name;
-        this.keys.addAll(keys);
+        for (final IKey key : keys) {
+            this.keys.add(SerializationUtils.clone(key));
+        }
     }
 
     public String getName() {
         return this.name;
     }
 
+    /**
+     * @return A list of all public keys of the contact
+     */
     public List<IKey> getKeys() {
-        return this.keys;
+        final ArrayList<IKey> keysCopy = new ArrayList<>();
+        for (final IKey key : this.keys) {
+            keysCopy.add(SerializationUtils.clone(key));
+        }
+        return keysCopy;
+    }
+
+    /**
+     * Adds a collection of keys to the public keys of the contact
+     * @param keys The public keys to add to the contact
+     */
+    public void addKeys(final Collection<IKey> keys) {
+        for (final IKey key : keys) {
+            this.keys.add(SerializationUtils.clone(key));
+        }
     }
 
     /**
      * {@inheritDoc}
      */
     @Override
-    public boolean equals(final Object o) {
-        if (o == null || this.getClass() != o.getClass()) {
+    public boolean equals(final Object obj) {
+        if (obj == null || this.getClass() != obj.getClass()) {
             return false;
         }
 
-        final Contact that = (Contact) o;
-        return this.name.equals(that.name)
-                && this.keys.equals(that.keys);
+        final Contact otherContact = (Contact) obj;
+        final List<IKey> otherKeys = otherContact.getKeys();
+        if (otherKeys.size() != this.keys.size()) {
+            return false;
+        }
+        for (int i = 0; i < this.keys.size(); i++) {
+            if (!this.keys.get(i).equals(otherKeys.get(i))) {
+                return false;
+            }
+        }
+        return this.name.equals(otherContact.getName());
```

Comment:
Because you were comparing a list with a list. If the other contact has a list with exactly the same size and contacts but with a different memory address, the old equals method would return false